### PR TITLE
feat(framework) Add `InMemoryDriver`

### PR DESCRIPTION
**Pull request by jafermarq**
This PR introduces a new type of driver that communicates in-memory (as opposed to via `gRPC`).

Then https://github.com/adap/flower/pull/3355 makes this `InMemoryDriver` the one to be used for simulations.

### CODE REVIEW

File: src/py/flwr/server/driver/inmemory_driver.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+        This method is used to collect messages from the SuperLink that correspond to a
+        set of given message IDs.
+        """
+        msg_ids = {UUID(msg_id) for msg_id in message_ids}
+        # Pull TaskRes
+        task_res_list = self.state.get_task_res(task_ids=msg_ids, limit=len(msg_ids))
+        # Convert TaskRes to Message
+        msgs = [message_from_taskres(taskres) for taskres in task_res_list]
+        return msgs
```
Comment (panh99) [inline src/py/flwr/server/driver/inmemory_driver.py:157]: I think we need to delete delivered TaskRes and the TaskIns they reply to after we get them from the `state`. Otherwise, the TaskIns and TaskRes will pile up in the `state.

**Review by panh99 — DISMISSED** [review]
lgtm!

File: src/py/flwr/server/driver/inmemory_driver.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+from .driver import Driver
+
+
+class InMemoryDriver(Driver):
+    """`InMemoryDriver` class provides an interface to the Driver API.
+
+    Parameters
+    ----------
+    state_factory : StateFactory
+        A StateFactory embedding a state designed for in-memory communication.
```
Comment (danieljanes) [inline src/py/flwr/server/driver/inmemory_driver.py:37]: Can't this use any kind of `State` implementation created by the `StateFactory`?

File: src/py/flwr/server/driver/inmemory_driver.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        self.fab_id = fab_id if fab_id is not None else ""
+        self.fab_version = fab_version if fab_version is not None else ""
+        self.node = Node(node_id=0, anonymous=True)
+        state = state_factory.state()
+
+        if not isinstance(state, InMemoryState):
+            raise NotImplementedError(
+                f"{self.__class__.__name__} is only compatible with an `InMemoryState` "
+                f"state. You used {type(state)}"
+            )
```
Comment (danieljanes) [inline src/py/flwr/server/driver/inmemory_driver.py:60]: Why is this necessary?

File: src/py/flwr/server/driver/inmemory_driver.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        return self.run_id
+
+    def create_message(  # pylint: disable=too-many-arguments
+        self,
+        content: RecordSet,
+        message_type: str,
+        dst_node_id: int,
+        group_id: str,
+        ttl: Optional[float] = None,
+    ) -> Message:
```
Comment (danieljanes) [inline src/py/flwr/server/driver/inmemory_driver.py:93]: Perhaps not in this PR, but in a later one: should we create a `BaseDriver` that has an implementation of `create_message` that is shared across all `Driver` implementations?

File: src/py/flwr/server/driver/inmemory_driver.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        """
+        task_ids: List[str] = []
+        for msg in messages:
+            # Check message
+            self._check_message(msg)
+            # Convert Message to TaskIns
+            taskins = message_to_taskins(msg)
+            # Store in state
+            taskins.task.pushed_at = time.time()
+            res = self.state.store_task_ins(taskins)
```
Comment (danieljanes) [inline src/py/flwr/server/driver/inmemory_driver.py:140]: Let's rename this for clarity:
```suggestion
            task_id = self.state.store_task_ins(taskins)
```

File: src/py/flwr/server/driver/inmemory_driver.py (hunk 3 above)
Comment (jafermarq) [reply src/py/flwr/server/driver/inmemory_driver.py]: removed

File: src/py/flwr/server/driver/inmemory_driver.py (hunk 2 above)
Comment (jafermarq) [reply src/py/flwr/server/driver/inmemory_driver.py:37]: yes

**Review by danieljanes — APPROVED** [review]
